Added realtime output for stdout and stderr

TheInsaneBro opened the pull request:
Since it's hard to know what the process is doing while it is running I've added two events to `Cli`. The events report when a new line gets written to either StdOut or StdErr. This will look like this then:

```c#
var ffmpeg = new Cli("ffmpeg.exe");
ffmpeg.StdErrDataReceived += (origin, data) => {
    // origin is the Cli-Object from which the event is coming from
    // data is the line written to StdErr (as string)
};
var output = await _ffmpeg.ExecuteAsync("args");
```

While it blocks when getting the output the event above will be called in case ffmpeg writes sth to the error stream.

I'm not sure if this is the best implementation in the async-method. If so, it could be refactored since it's very similar to the synchronous way of doing it.

pipe01 commented:
I suggest using an EventArgs class for both of the events and changing them to (object sender, StdEventArgs e).

TheInsaneBro commented:
Changed it

Tyrrrz commented:
I don't really like this implementation, because it's very common to run many executions at the same time, even the sync methods (which can be ran on different threads).

Besides, the idea of CliWrap is to abstract away from working with process and buffer. A typical CLI command is very much like a method you can call in a library, and CliWrap tries to present it as such, so I'm not entirely sure having those events is appropriate.

TheInsaneBro commented:
The reason I've implemented this is that I can get a progress update from ffmpeg. If I render a video that is a few hours long, I want to see how's the progress at times.

My first thought was using a callback as argument in `ExecuteAsync(...)`, but with all the overloads it's hard to implement smoothly. But that would solve the problem of parallel executions.

If we decide not to implement it I have to fall back to use .Net's Processes which I really dislike. (actually I would just use the fork, but I might not be the only one with that concern)

Tyrrrz commented:
Hmm. I can see why you would need to use it with Ffmpeg.

I think the correct approach here would be to make an interface similar to `IProgress`, call it `IStandardBufferHandler` that has `void HandleStandardOutput(string line)` and `void HandleStandardError(string line)`. And then make a class that allows setting those from delegates. It's important that the handler is a method parameter because the Cli class itself doesn't contain any state.

Tyrrrz commented:
I want to include this feature in the next version. @TheInsaneBro do you have time to finish it or should I?

TheInsaneBro commented:
When is the next version planned? I could finish it in 2 - 3 days.

Tyrrrz commented:
It's planned for whenever this is merged. :p

TheInsaneBro commented:
I think I can comply to that

TheInsaneBro commented:
Added it as you suggested.

I have changed the `Execute` and `ExecuteAsync`-Calls a bit to not have a gazillion overloads. Also added some `#region` to Cli.cs.

Lastly I've created a standard implementation of `IStandardBufferHandler` since you referred to `IProgress<T>` and there is one for it too (`Progress<T>`)

Tyrrrz commented:
Thank you!
I will try to look at this tomorrow.

Tyrrrz commented on CliWrap/Cli.cs:
```diff
@@ -27,7 +27,7 @@ public class Cli
         /// </summary>
         public string WorkingDirectory { get; }
-
+        
```
Please remove this rogue whitespace :p

Tyrrrz commented on CliWrap/Cli.cs:
```diff
@@ -94,13 +96,18 @@ public ExecutionOutput Execute(ExecutionInput input, CancellationToken cancellat
                 process.OutputDataReceived += (sender, args) =>
                 {
-                    if (args.Data != null)
+                    if (args.Data != null) {
```
Can you make the opening braces start on new line please (here and in other places)?

Tyrrrz commented on CliWrap/Cli.cs:
```diff
@@ -255,33 +271,23 @@ public void ExecuteAndForget()
         /// <summary>
         /// Executes CLI without input, waits until completion asynchronously and returns output
         /// </summary>
         public Task<ExecutionOutput> ExecuteAsync()
```
You removed the argument-less Execute but kept one for ExecuteAsync. I suggest to remove this one too to make it consistent.

Tyrrrz commented on CliWrap/StandardBufferHandler.cs:
```diff
@@ -0,0 +1,40 @@
+        /// <summary>
+        /// Sets up a standard buffer handler using two callbacks. Both callbacks are optional.
+        /// </summary>
+        /// <param name="standardOutputHandler">The callback to use when it gets written to the standard output</param>
```
`The callback to use when it gets written...`
maybe change to `The callback to use when a line is written...` to make it more clear

Tyrrrz commented on CliWrap/StandardBufferHandler.cs:
```diff
@@ -0,0 +1,40 @@
+namespace CliWrap
+{
+    /// <summary>
+    /// The standard implementation of <see cref="IStandardBufferHandler"/>
```
Let's call it `the default implementation` to avoid using word "standard" in two different meanings.

Tyrrrz commented on CliWrap/StandardBufferHandler.cs:
```diff
@@ -0,0 +1,40 @@
+        }
+
+        /// <summary>
+        /// Calls the callback
```
You can just use `/// <inheritdoc />` on the following two methods to have them inherit comments from the interface

TheInsaneBro commented:
Should be alright now

Tyrrrz commented:
Great

TheInsaneBro commented:
Nice, thanks for taking that :+1:

Tyrrrz commented:
Already on nuget, should be indexed soon so you can update
